### ISSUE TITLE
Internal error occured: TypeError: Cannot read properties of undefined (reading 'name')

### ISSUE THREAD

**Issue by Daraan** [microsoft/pyright#8757]
**Describe the bug**

I am getting the following error at the top of one of my files:
`An internal error occurred while type checking file "/home/username/.../classes/rule.py": TypeError: Cannot read properties of undefined (reading 'name')`; full traceback at the end.

**Version and Setup**

- Command Line: v1.1.376

- pyright from pylance in VSCode
  PyLance:
  - PreRelease v2024.8.101: pyright version: `1.1.375`, commit d2c0938f
  - Release v2024.8.1: pyright version `1.1.373`, commit ee424479

**Code**

Below is a minimal working example.
The error comes from a seemingly overloaded / inconsistent `__call__` method.

The `ConditionFunction` class that I am using is a descriptor protocol. It should be able to decorate functions(ctx), methods(rule, ctx) and external functions(rule, ctx) that have the same signature as a method.
This creates some duality of one or two positional arguments in the `__call__` method and I guess the ParamSpec struggles there.

```python
from __future__ import annotations

from typing import (TypeVar, Union, Callable, Hashable)
from typing_extensions import (TypeAlias, Self, ParamSpec)

from functools import partial

from typing import Generic, Hashable, TYPE_CHECKING, Optional, TypeVar, cast, Any
from typing_extensions import overload, Self, ParamSpec

if TYPE_CHECKING:
    from classes.rule import Context, Rule # circular import

_CH = TypeVar("_CH", bound=Hashable) # Generic of ConditionFunction
_CP = ParamSpec("_CP") # Generic of ConditionFunction
_Rule = TypeVar("_Rule", bound="Rule")

class ConditionFunction(Generic[_Rule, _CP, _CH]):
    
    @overload
    def __call__(self, ctx: "Context", *args:_CP.args, **kwargs:_CP.kwargs) -> _CH: ...
    # Function call
    
    @overload
    def __call__(self, rule: "Rule", ctx: "Context", *args:_CP.args, **kwargs:_CP.kwargs) -> _CH: ...
    # Method call

    def __call__(self, ctx: "Rule | Context", *args:_CP.args, **kwargs:_CP.kwargs) -> _CH:
       # In case of Method call ctx is Rule and args[0] is Context.
        return cast(_CH, 0)
    
    # Error consists when this is removed as well but this shows the use case
    def __get__(self, instance: Optional[Rule], objtype: Optional["type[Rule]"]=None) -> "Self | partial[_CH]":
        """
        :term:`Descriptor Protocol <descriptor>`, for in class usage like Rule.condition
        """
        # NOTE: instance.condition is not an ConditionFunction, it is a partial of one.
        if instance is None:
            return self # called on class Rule.condition
        return partial(self, instance) # NOTE: This fixes "ctx" to instance in __call__, the real "ctx" in __call__ is provided through *args
    

_T = TypeVar("_T")
_H = TypeVar("_H", bound=Hashable)
_P = ParamSpec("_P")

_Rule = TypeVar("_Rule", bound="Rule")

ConditionFunctionLike : TypeAlias = Union[ConditionFunction[_Rule, _P, _H], # XXX
                                          Callable[["Context"], _H], 
                                          Callable[[_Rule, "Context"], _H]]

OkUnion1  : TypeAlias = Union[Callable[[_Rule, "Context"], _H],
                             Callable[["Context"], _H], 
                             ]

OkUnion2 : TypeAlias = Union[ConditionFunction[_Rule, ..., _H], # XXX
                                          Callable[["Context"], _H], 
                                          Callable[[_Rule, "Context"], _H]]

class Context: ...

class Rule():
   
    condition: ConditionFunction[Self, _P, Hashable] # <--- XXX error version
    #condition: ConditionFunctionLike[Self, ..., Hashable] # <--- XXX OK
    #condition: OkUnion[Self, Hashable]  # <--- XXX OK
    #condition: OkUnion2[Self, Hashable]  # <--- XXX OK

    def evaluate(self, ctx: Context):
        reveal_type(self.condition)
        result = self.condition(ctx)   # <--- XXX error version
        return result

```

---

<details>
<summary>Traceback - VSCode pyright version: `1.1.375` </summary>

**Traceback - VSCode**
```
An internal error occurred while type checking file "/home/username/.../classes/rule.py": TypeError: Cannot read properties of undefined (reading 'name')
    at on (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:712453)
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:697494
    at Array.forEach (<anonymous>)
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:697473
    at Za (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:832479)
    at tn (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:697409)
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:701464
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:702013
    at Za.allowDiagnostics (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:708541)
    at Za (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:832432)
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:699986
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:649342
    at y (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:966575)
    at s (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:648708)
    at xt (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:649811)
    at an (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:699948)
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:707973
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:708294
    at Za.allowDiagnostics (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:708541)
    at Za (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:832479)
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:699986
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:649342
    at y (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:966575)
    at s (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:648708)
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:649769
    at Array.forEach (<anonymous>)
    at xt (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:649752)
    at an (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:699948)
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:611842
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:612476
    at $a (/home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:832204)
    at /home/username/.vscode-server/extensions/ms-python.vscode-pylance-2024.8.101/dist/pyright.bundle.js:1:604867Pylance
```

</details>

<details>
<summary>Traceback CommandLine v1.1.376</summary>
**Traceback CommandLine v1.1.376**

```
An internal error occurred while type checking file "/home/username/TeamProject/LunaticAI-Driver-for-CARLA-Simulator/classes/rule.py": TypeError: Cannot read properties of undefined (reading 'name')
    at matchArgsToParams (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:10349:74)
    at forEach (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9086:38)
    at Array.forEach (<anonymous>)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9082:47)
    at useSpeculativeMode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:20833:28)
    at validateOverloadedArgTypes (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9080:9)
    at validateCallForOverloaded (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9719:28)
    at validateCallArgsForSubtype (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9457:24)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9364:44)
    at useSpeculativeMode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:20827:20)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9361:24)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:3981:43)
    at doForEachSubtype (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeUtils.ts:802:9)
    at expandSubtype (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:3965:13)
    at mapSubtypesExpandTypeVars (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:4023:13)
    at validateCallArgs (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9357:26)
    at validateCallForClassInstance (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:10099:28)
    at validateCallArgsForSubtype (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9486:24)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9364:44)
    at useSpeculativeMode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:20833:28)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9361:24)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:3981:43)
    at doForEachSubtype (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeUtils.ts:802:9)
    at expandSubtype (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:3965:13)
    at forEach (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:4020:17)
    at Array.forEach (<anonymous>)
    at mapSubtypesExpandTypeVars (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:4019:22)
    at validateCallArgs (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9357:26)
    at getTypeOfCall (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:8110:36)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:1180:62)
    at useSignatureTracker (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:20786:28)
    at getTypeOfExpressionCore (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:1180:30)
    at getTypeOfExpression (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:1085:26)
    at evaluateTypesForAssignmentStatement (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:16206:35)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:19779:25)
    at a.evaluateTypesForStatement [as timeOperation] (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/common/timing.ts:40:20)
    at K.visitAssignment (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:1180:25)
    at K.visit (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:367:29)
    at K.visitNode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:933:21)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:915:37)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:285:19)
    at forEach (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:924:22)
    at Array.forEach (<anonymous>)
    at K.walkMultiple (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:922:15)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:917:18)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:285:19)
    at K._walkStatementsAndReportUnreachable (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:2880:18)
    at K.visitSuite (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:294:14)
    at K.visit (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:544:29)
    at K.visitNode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:933:21)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:915:37)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:285:19)
    at K.visitFunction (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:663:18)
    at K.visit (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:451:29)
    at K.visitNode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:933:21)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:915:37)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:285:19)
    at K._walkStatementsAndReportUnreachable (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:2880:18)
    at K.visitSuite (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:294:14)
    at K.visit (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:544:29)
    at K.visitNode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:933:21)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:915:37)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:285:19)
    at K.visitClass (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:319:14)
Error performing analysis: TypeError: Cannot read properties of undefined (reading 'name')
    at matchArgsToParams (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:10349:74)
    at forEach (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9086:38)
    at Array.forEach (<anonymous>)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9082:47)
    at useSpeculativeMode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:20833:28)
    at validateOverloadedArgTypes (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9080:9)
    at validateCallForOverloaded (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9719:28)
    at validateCallArgsForSubtype (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9457:24)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9364:44)
    at useSpeculativeMode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:20827:20)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9361:24)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:3981:43)
    at doForEachSubtype (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeUtils.ts:802:9)
    at expandSubtype (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:3965:13)
    at mapSubtypesExpandTypeVars (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:4023:13)
    at validateCallArgs (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9357:26)
    at validateCallForClassInstance (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:10099:28)
    at validateCallArgsForSubtype (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9486:24)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9364:44)
    at useSpeculativeMode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:20833:28)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9361:24)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:3981:43)
    at doForEachSubtype (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeUtils.ts:802:9)
    at expandSubtype (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:3965:13)
    at forEach (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:4020:17)
    at Array.forEach (<anonymous>)
    at mapSubtypesExpandTypeVars (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:4019:22)
    at validateCallArgs (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:9357:26)
    at getTypeOfCall (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:8110:36)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:1180:62)
    at useSignatureTracker (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:20786:28)
    at getTypeOfExpressionCore (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:1180:30)
    at getTypeOfExpression (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:1085:26)
    at evaluateTypesForAssignmentStatement (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:16206:35)
    at callback (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:19779:25)
    at a.evaluateTypesForStatement [as timeOperation] (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/common/timing.ts:40:20)
    at K.visitAssignment (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:1180:25)
    at K.visit (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:367:29)
    at K.visitNode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:933:21)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:915:37)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:285:19)
    at forEach (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:924:22)
    at Array.forEach (<anonymous>)
    at K.walkMultiple (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:922:15)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:917:18)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:285:19)
    at K._walkStatementsAndReportUnreachable (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:2880:18)
    at K.visitSuite (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:294:14)
    at K.visit (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:544:29)
    at K.visitNode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:933:21)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:915:37)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:285:19)
    at K.visitFunction (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:663:18)
    at K.visit (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:451:29)
    at K.visitNode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:933:21)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:915:37)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:285:19)
    at K._walkStatementsAndReportUnreachable (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:2880:18)
    at K.visitSuite (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:294:14)
    at K.visit (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:544:29)
    at K.visitNode (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:933:21)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:915:37)
    at K.walk (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:285:19)
    at K.visitClass (/home/username/.cache/pyright-python/1.1.376/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:319:14)
```
</details>

---

I try to further track down the error but I am a bit stuck what to look out for. Tell me if I can help you with anything.

**Comment by erictraut** [comment]
Thanks for the bug report. Unfortunately, I haven't been able to repro the issue. Without a repro, I don't think I'm going to be able to find the root cause of the problem. I cloned your repo and tried to repro the issue using the full source base, but I'm unable to successfully pip install `carla` on my machine because it apparently doesn't support the Apple silicon processor architecture. 

Based on the stack trace you provided, I've narrowed down the crash to a particular statement on a particular line, but I haven't been able to determine how to recreate the conditions for the crash. I could add some additional defensive checks to avoid the crash, but this would likely move the crash elsewhere. So I'm going to refrain from doing this and wait until I can find a repro case.

Any additional effort toward finding a minimal, self-contained repro would be really helpful. One technique I sometimes use is to bisect a source file by commenting out half of the functions/classes/methods in a source file to see whether the crash still repros. By continually bisecting, it's usually possible to narrow the issue to a specific function and even a specific line within a function. From that point, it's usually pretty straightforward to extract the code and its dependencies to make a stand-alone sample that repros the issue.

**Comment by Daraan** [comment]
Thank you for the reply and for looking into it.

For the meantime I added my stub for carla file to the repo in a new branch:
https://github.com/Daraan/LunaticAI-Driver-for-CARLA-Simulator/tree/_pyright_error

In the that example the I can switch the error on and off with the lines [L605-L606](https://github.com/Daraan/LunaticAI-Driver-for-CARLA-Simulator/blob/6aeecc20f3c389c014956e96a8d13245d811f012/classes/rule.py#L605-L606).

**Comment by Daraan** [comment]
EDIT2:

Minimal working example added that can reproduce the error added.

**Comment by erictraut** [comment]
Thanks! That's really helpful. With your assistance, I was able to repro the problem, identify the root cause, and introduce a bug fix. This issue will be addressed in the next release.

The reason I haven't seen this before is that the bug can be triggered only if you use a ParamSpec in a way that is not permitted by the typing spec. The ParamSpec `_P` in the statement `condition: ConditionFunction[Self, _P, Hashable]` is not associated with any scope and therefore cannot be used in this manner. Pyright correctly flags this as an error, but this condition later leads to an unexpected situation that the code was not properly handling.

Here's a more minimal repro case.

```python
from typing import Generic, ParamSpec

P = ParamSpec("P")

class A(Generic[P]):
    def __call__(self, a: int, b: int, *args: P.args, **kwargs: P.kwargs) -> None:
        ...

class B:
    x: A[P]  # Illegal use of ParamSpec

B().x(1)  # Causes pyright to crash
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.377.
